src/main/scala/com/databricks/spark/xml/functions.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2019 Databricks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.databricks.spark.xml
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.{Column, Dataset, Row}
+import org.apache.spark.sql.catalyst.parser.CatalystSqlParser
+import org.apache.spark.sql.types.{ArrayType, DataType, StructType}
+
+import com.databricks.spark.xml.parsers.StaxXmlParser
+import com.databricks.spark.xml.util.InferSchema
+
+/**
+ * Support functions for working with XML columns directly.
+ */
+object functions {
+
+  /**
+   * Infers the schema of XML documents as strings.
+   *
+   * @param ds Dataset of XML strings
+   * @param options additional XML parsing options
+   * @return inferred schema for XML
+   */
+  @Experimental
+  def schema_of_xml(ds: Dataset[String], options: Map[String, String] = Map.empty): StructType =
+    InferSchema.infer(ds.rdd, XmlOptions(options))
+
+  /**
+   * Infers the schema of XML documents when inputs are arrays of strings, each an XML doc.
+   *
+   * @param ds Dataset of XML strings
+   * @param options additional XML parsing options
+   * @return inferred schema for XML. Will be an ArrayType[StructType].
+   */
+  @Experimental
+  def schema_of_xml_array(ds: Dataset[Array[String]],
```

Comment:
Also, new method for getting the ArrayType schema out of an Array of strings. Not super pretty but helpful

Reply:
Hm, @srowen, I think this can be elegantly worked around via:

```scala
df.select(transform(col("array_of_xml"), x => from_xml(x)))
``` 

Although those were added as of Spark 3 (SPARK-27297).

e.g.)

```scala
scala> spark.range(1).selectExpr("array(1,2,3) as arr").select(transform(col("arr"), x => concat(x))).printSchema()
root
 |-- transform(arr, lambdafunction(concat(x), x)): array (nullable = false)
 |    |-- element: string (containsNull = false)
```

Reply:
Hm, so can this work in Spark 2.x then? we'd still have to support it.
I agree that's much more elegant for Spark 3 though

Reply:
Okay .. I think we can deprecate this behaviour later when Spark XML starts to deprecate Spark 2 supports.